src/EventStore.Core/Services/TimerService/ThreadBasedScheduler.cs

```diff
@@ -90,9 +89,23 @@ public class ThreadBasedScheduler : IMonitoredQueue, IScheduler, IDisposable {
 					_queueStats.ProcessingEnded(processed);
 
 					if (processed == 0) {
-						_queueStats.EnterIdle();
+						var timeout = TimeSpan.FromSeconds(30);
+
+						if (_tasks.Count > 0) {
+							var ticksUntilNextDueTime = _tasks.FindMin().DueTime.ElapsedTicksSince(_timeProvider.Now);
+							var secondsUntilNextDueTime = (double)ticksUntilNextDueTime / Instant.TicksPerSecond;
+
+							if (secondsUntilNextDueTime <= 0)
+								continue;
```

Comment:
```suggestion
							var secondsUntilNextDueTime = (double)ticksUntilNextDueTime / Instant.TicksPerSecond;

							if (secondsUntilNextDueTime <= 0)
								continue;
```

@taspeotis it seems possible to get rid of the `secondsUntilNextDueTime` variable and do the `<= 0` check with `ticksUntilNextDueTime` itself. `dueTimeSpan` can then be built with `TimeSpan.FromTicks()` instead of `TimeSpan.FromSeconds`

Reply:
this would save us from doing any division operation / any approximation

Reply:
sorry, i missed the fact that we're working with `Instant`s here not `TimeSpan`s, so we won't be able to use `TimeSpan.FromTicks`.  Maybe we could try to add an operation in `Instant` to convert it to a `TimeSpan`

Reply:
Hi, please give me a day or two to play with it. Thank you for the review and feedback!

Also if you have the time to do it, I think I enabled "contributors can make changes" in GitHub. I've never used it before but I think you can push to my branch.

Reply:
thanks for the quick response @taspeotis yup sure, i'll give it a try, it shouldn't be a big change :+1: 
by any luck, do you also have some numbers/profiler runs after the fix?

Reply:
i've pushed a commit with what i had in mind, feel free to let me know if you find the change appropriate (i've not yet thoroughly tested it though)

i've also adjusted the max. timeout to 5 seconds instead of 30 seconds as in practice, it looks like the actual timeout is of the order of a few hundred milliseconds (and also just in case there's any bug, we know that the task will be processed within 5 seconds max)

Reply:
Thanks @shaan1337 I'll try and benchmark this shortly

Reply:
thanks! feel free to push any additional commits if required